index.bs

```diff
@@ -698,9 +698,23 @@ Native hit test result {#native-hit-test-result-section}
 
-The information about position and orientation of the intersection point should be contained in [=native hit test result=]'s native origin. Native origin defines a new coordinate system in such a way that its Y axis represents the surface's normal vector at the intersection point. If the orientation is not returned from the XR device, the user agent SHOULD set the native origin in such a way that Y axis of the coordinate system it defines is pointing up (towards negative gravity vector).
+The information about position and orientation of the intersection point should be contained in [=native hit test result=]'s native origin. Native origin defines a new coordinate system in such a way that its Y axis represents the surface's normal vector at the intersection point. If the orientation is not returned from the XR device, the user agent SHOULD set the native origin in such a way that Y axis of the coordinate system it defines is set to the normalized, negated gravity vector. Additionally, if possible, the Z axis of the returned native origin should be set so that it points towards the origin of the ray used to perform the native hit test (i.e. the ray's origin lies on the YZ plane). The exact steps to obtain the native origin are described in [=calculate the native origin=] algorithm.
 
-Issue: Decide if we need to specify other axes of the coordinate system defined by hit test result's native origin to maintain compatibility between different implementations & differrent AR frameworks.
+<div class="algorithm" data-algorithm="calculate-native-origin">
+
+To <dfn>calculate the native origin</dfn> given [=native hit test result=] |result|, {{XRRay}} |offsetRay| and a coordinate system |coordinates|, the user agent MUST run the following steps:
+    1. Let |normal| be the surface normal of the |result| expressed in coordinate system |coordinates|. If the native hit test result does not return a surface normal, set |normal| to a negative gravity vector & normalize.
+    1. Let |p| be the intersection point of the |result| expressed in coordinate system |coordinates|.
+    1. Let |origin| be |offsetRay|'s {{XRRay/origin}}.
+    1. Let |Z| be a directional vector resulting from normalizing |origin| - |p|.
+    1. If |Z| and |normal| are parallel (i.e. they are supported by parallel lines), set |Z| to <code>{ x: 0.0, y: 1.0, z: 00, w: 0.0 }</code>.
+    1. If |Z| and |normal| are parallel, set |Z| to <code>{ x: 1.0, y: 0.0, z: 0.0, w: 0.0 }</code>.
+    1. Let |X| be a directonal vector resutling from normalizing |normal| x |Z|.
+    1. Let |rotation| be a rotation that maps |X| to directional vector <code>{ x: 1.0, y: 0.0, z: 0.0, w: 0.0 }</code>, |normal| to directional vector <code>{ x: 0.0, y: 1.0, z: 0.0, w: 0.0 }</code>, and |Z| to directonal vector <code>{ x: 0.0, y: 0.0, z: 1.0, w: 0.0 }</code>.
```

Comment:
This is underspecified and probably should be mathed out

Reply:
Actually, hmm, we don't do the math in any part of the spec, and instead use wording like 

> Set transform’s [orientation](https://immersive-web.github.io/webxr/#dom-xrrigidtransform-orientation) to the orientation of space’s [effective origin](https://immersive-web.github.io/webxr/#xrspace-effective-origin) in baseSpace’s [coordinate system](https://immersive-web.github.io/webxr/#xrspace-coordinate-system).

Perhaps stick to that kind of wording?

Reply:
> This is underspecified and probably should be mathed out

My understanding is that it's not underspecified for our purposes? We need a specific orientation of the newly created pose and do not really care what rotation was used to achieve it, as long as the end result is as described.

> (...) Perhaps stick to that kind of wording?

The problem I have here is that I don't have an effective origin. I know the axes of the coordinate system when expressed in other space's coordinate system (|X|, |normal|, and |Z|, expressed relative to |coordinate system|), and all we need here is a coordinate system change matrix ("just smack those vectors in a matrix and you're done"), but I don't really see how to express it differently. 😕

Reply:
Ah, fair.